transport/src/main/java/io/netty/channel/ChannelFactoriesRegistry.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * This class provides a default mapping between {@link EventLoopGroup} implementations and {@link ChannelFactory}
+ * implementations
+ */
+public abstract class ChannelFactoriesRegistry {
+
+    private static final class ChannelFactoryDefaults {
+        private final Class<? extends EventLoopGroup> eventLoopGroupType;
+        private final ChannelFactory clientSideFactory;
+        private final ChannelFactory serverSideFactory;
+
+        public ChannelFactoryDefaults(Class<? extends EventLoopGroup> eventLoopGroupType,
+                                      ChannelFactory clientSideFactory, ChannelFactory serverSideFactory) {
+            this.eventLoopGroupType = eventLoopGroupType;
+            this.clientSideFactory = clientSideFactory;
+            this.serverSideFactory = serverSideFactory;
+        }
+
+        private boolean matchesClass(Class<? extends EventLoopGroup> clazz) {
+            return eventLoopGroupType.isAssignableFrom(clazz);
+        }
+    }
+
+    private static final List<ChannelFactoryDefaults> DEFAULT_CHANNEL_FACTORIES
```

Comment:
I think a COW would make more sense here so we can get rid of the sync.

Reply:
I think we can't. Otherwise we can't implement correctly registerFactoryForEventLoopGroup.